### PR TITLE
Fix sub menus don't work with IE11/Edge

### PR DESCRIPTION
**Pull request by benoitchantre**
## Description

Addresses issue #437.

Avoids the usage of a polyfill by replacing `forEach` loops by `for` loops.
Even it prevents previous errors caused by `forEach` on IE11, sub menus are not working.

## List of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] This pull request relates to a ticket.
- [ ] This code is tested.
- [ ] This change has been added to CHANGELOG.md
- [ ] I want my code added to WP Rig.

### CODE REVIEW

**Comment by benoitchantre** [comment]
@mor10 There was indeed a CSS issue.

**Comment by benoitchantre** [comment]
Note (to check with a test): when Javascript is disabled, sub menus will not be displayed for browsers that don't [support `:focus-within`](https://caniuse.com/#search=focus-within). If it has to be fixed, this could be done in a follow-up PR.

**Comment by benoitchantre** [comment]
@mor10 Can you milestone that for `v2.0`

File: assets/js/src/navigation.js [hunk 1]
```diff
@@ -38,9 +38,9 @@ function initNavToggleSubmenus() {
 		return;
 	}
 
-	navTOGGLE.forEach( function( nav ) {
-		initEachNavToggleSubmenu( nav );
-	});
+	for ( let i = 0; i < navTOGGLE.length; i++ ) {
+		initEachNavToggleSubmenu( navTOGGLE[ i ]);
```
Comment (bamadesigner) [inline assets/js/src/navigation.js]: Can you fix the spacing to be `initEachNavToggleSubmenu( navTOGGLE[i] );`?

File: assets/js/src/navigation.js [hunk 2]
```diff
@@ -76,7 +76,7 @@ function initEachNavToggleSubmenu( nav ) {
 			dropdown.appendChild( dropdownSymbol );
 
 			// Add before submenu.
-			submenu.parentNode.insertBefore( dropdown, submenu );
+			SUBMENUS[ i ].parentNode.insertBefore( dropdown, SUBMENUS[ i ]);
```
Comment (bamadesigner) [inline assets/js/src/navigation.js]: Can you fix the spacing? `SUBMENUS[i].parentNode.insertBefore( dropdown, SUBMENUS[i] );`

File: assets/js/src/navigation.js [hunk 3]
```diff
@@ -147,9 +148,9 @@ function initNavToggleSmall() {
 		return;
 	}
 
-	navTOGGLE.forEach( function( nav ) {
-		initEachNavToggleSmall( nav );
-	});
+	for ( let i = 0; i < navTOGGLE.length; i++ ) {
+		initEachNavToggleSmall( navTOGGLE[ i ]);
```
Comment (bamadesigner) [inline assets/js/src/navigation.js]: Can you fix the spacing? `initEachNavToggleSmall( navTOGGLE[i] );`

File: assets/js/src/navigation.js [hunk 4]
```diff
@@ -129,9 +130,9 @@ function initEachNavToggleSubmenu( nav ) {
 			}
 		});
 
-		submenu.parentNode.classList.add( 'menu-item--has-toggle' );
+		SUBMENUS[ i ].parentNode.classList.add( 'menu-item--has-toggle' );
```
Comment (bamadesigner) [inline assets/js/src/navigation.js]: Can we make all array usage match? `SUBMENUS[i]`

File: assets/js/src/navigation.js [hunk 5]
```diff
@@ -203,16 +204,19 @@ function toggleSubMenu( parentMenuItem, forceToggle ) {
...
 
 		// Make sure all children are closed.
-		parentMenuItem.querySelectorAll( '.menu-item--toggled-on' ).forEach( function( item ) {
-			toggleSubMenu( item, false );
-        });
+		const subMenuItemsToggled = parentMenuItem.querySelectorAll( '.menu-item--toggled-on' );
+		for ( let i = 0; i < subMenuItemsToggled.length; i++ ) {
+			toggleSubMenu( subMenuItemsToggled[ i ], false );
+		}
+
```
Comment (bamadesigner) [inline assets/js/src/navigation.js]: Please remove this extra blank line.

File: assets/js/src/navigation.js (hunk 1 above)
Comment (benoitchantre) [reply assets/js/src/navigation.js]: Done, but the build script reports a syntax error (`42:42  error  There should be no spaces inside this paren  space-in-parens`). Any idea why? Cc @ataylorme.

File: assets/js/src/navigation.js (hunk 2 above)
Comment (benoitchantre) [reply assets/js/src/navigation.js]: Done, but similar error reported by the build script.

File: assets/js/src/navigation.js (hunk 3 above)
Comment (benoitchantre) [reply assets/js/src/navigation.js]: Done, but similar error reported by the build script.

File: assets/js/src/navigation.js (hunk 4 above)
Comment (benoitchantre) [reply assets/js/src/navigation.js]: Done.

File: assets/js/src/navigation.js (hunk 5 above)
Comment (benoitchantre) [reply assets/js/src/navigation.js]: Done.

**Comment by ataylorme** [comment]
@benoitchantre the errors are coming from ESLint under the [`space-in-parens` rule](https://eslint.org/docs/rules/space-in-parens). [ESLint is configured to use the WordPress coding standards](https://github.com/wprig/wprig/blob/v2.0/package.json#L79) via [`eslint-config-wordpress`](https://www.npmjs.com/package/eslint-config-wordpress).

However, that package has not been updated in 2 years so may not reflect current coding standards. WordPress core uses JSHint with a [`.jshintrc` file in trunk](https://develop.svn.wordpress.org/trunk/.jshintrc). Perhaps we should switch to JSHint as well or use a tool like [polyjuice](https://github.com/brenolf/polyjuice) to do a conversion of the core `.jshintrc` file for ESLint.

I think re-working that functionality is a new issue.

**Comment by benoitchantre** [comment]
Thank you @ataylorme for the detailed explanation.

**Comment by ataylorme** [comment]
> No spaces between [i] and add spaces between function calls.

It could also be that the package we use is correctly checking the rules and what @bamadesigner is requesting is in violation of the coding standard

**Comment by benoitchantre** [comment]
Here's some tests on line 42 with the current build script:

❌ `initEachNavToggleSubmenu( navTOGGLE[ i ] );`
✅ `initEachNavToggleSubmenu( navTOGGLE[ i ]);`
❌ `initEachNavToggleSubmenu( navTOGGLE[i] );`
✅ `initEachNavToggleSubmenu( navTOGGLE[i]);`

**Comment by bamadesigner** [comment]
I mean `[i]` vs `[ i ]` I can understand my personal preference going up against a "standard" but `( test)` being preferred over `( test )` has to be a bug.

**Comment by ataylorme** [comment]
I think the source of the problem is that we are using the [`eslint-config-wordpress`](https://www.npmjs.com/package/eslint-config-wordpress) to get the ESLint config but it hasn't been updated in 2 years.

**Comment by ataylorme** [comment]
@benoitchantre I think you can set the code style to how @bamadesigner prefers so we can get this merged. I opened #528 for the larger issues of the ESLint config being broken/outdated.

**Comment by benoitchantre** [comment]
@ataylorme done in https://github.com/wprig/wprig/pull/438/commits/e02f544ca64ba81cd72c9033f0193649edb11153

**Pull request merged**